integration/commands_authentication_test.go

```diff
@@ -15,9 +15,110 @@
 package integration
 
 import (
+	"strings"
 	"testing"
+
+	"github.com/stretchr/testify/require"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+
+	"github.com/FerretDB/FerretDB/integration/setup"
 )
 
 func TestCommandsAuthenticationSASLStart(t *testing.T) {
-	// TODO https://github.com/FerretDB/FerretDB/issues/1568
+	setup.SkipForTigrisWithReason(t, "postgreSQL authentication test")
+
+	t.Parallel()
+
+	testcases := map[string]struct {
+		skip            string
+		dbErr           string
+		listenerTLS     bool
+		clientTLS       bool
+		invalidUsername bool
+		invalidPassword bool
+	}{
+		"TLS": {
+			listenerTLS: true,
+			clientTLS:   true,
+		},
+		"TCP": {
+			listenerTLS: false,
+			clientTLS:   false,
+		},
+		"TLSWrongUsername": {
+			listenerTLS:     true,
+			clientTLS:       true,
+			invalidUsername: true,
+			dbErr:           "role \"invalid\" does not exist",
+		},
+		"TLSWrongPassword": {
+			listenerTLS:     true,
+			clientTLS:       true,
+			invalidPassword: true,
+			skip:            "TODO: Expects error but connection is established successfully",
+		},
+		"TLSWithNonTLSClient": {
+			listenerTLS: true,
+			clientTLS:   false,
+			skip:        "MongoDB driver attempts to reconnect until time out",
+		},
+	}
+
+	for name, tc := range testcases {
+		name, tc := name, tc
+		t.Run(name, func(t *testing.T) {
+			if tc.skip != "" {
+				setup.SkipForPostgresWithReason(t, tc.skip)
+			}
+
+			postgreSQLURL := "postgres://username:password@127.0.0.1:5432/ferretdb?pool_min_conns=1"
+			port := 0
+			unix := false
+			s := setup.SetupWithOpts(t, &setup.SetupOpts{
+				Flags: map[string]any{
+					"target-tls":         tc.listenerTLS,
+					"target-port":        port,
+					"target-unix-socket": unix,
+					"postgresql-url":     postgreSQLURL,
+				},
+			})
+
+			ctx := s.Ctx
+
+			// s.MongoDBURI looks like `mongodb://username:password@127.0.0.1:35697/?authMechanism=PLAIN`.
+			// For testing invalid parameter, we replace each part of it.
+			clientURI := s.MongoDBURI
+
+			if tc.invalidUsername {
+				clientURI = strings.Replace(clientURI, "username", "invalid", 1)
```

Comment:
I believe the safer way to do that is to use net/url, although it's just a test, so I believe it's okay

Reply:
Thanks good idea, I ended up using options package from the driver to set it :hugs:

Reply:
:100: